DS-2635 discovery facets with configurable ordering

jonas-atmire opened the pull request:
This contribution adds the possibility to differentiate the standard sort orders for discovery.
Initially, upon clicking the "view more" link in a discovery facet, the user is simply sent to an alphabetically sorted page.
It would be more logical that, when the facet is sorted based on the count, the view more page isn't hard-set to sort on this as well, but rather be able to have its own configurable sorting. 
A differentiation between the discovery facets and actual view-more pages has been added to be able to configure them through the discovery.xml file. This allows for a more custom configuration for different facets. 
This fix is based on the following JIRA ticket - https://jira.duraspace.org/browse/DS-2635
Atmire (http://www.atmire.com) has been commissioned by CGIAR (http://www.cgiar.org/) to fix this particular JIRA ticket as wel as helping improve the usability of DSpace by contributing it to the community.

jonas-atmire committed:
DS-2635 Configurable order by value/count per discovery facet

Modifications to the following files:
-searchFacetFilter.java
-SidebarFacetsTransformer.java
,enable the user to change what kind of sort order is used depending on what discovery facet is clicked

The was an improvement based on the following Jira ticket
https://jira.duraspace.org/browse/DS-2635

jonas-atmire committed:
DS-2635 Configuration of sort order split into 2 parts

Originally, one sortOrder configured in the discovery.xml spring file would be used.
This had an impact on both the sidebar and view more pages
This commit splits this sortOrder to make it possible to have different sortOrder for the view more page and sidebar facet

This was an improvement based on the following Jira ticket
https://jira.duraspace.org/browse/DS-2635

jonas-atmire committed:
DS-2635 Discovery facets splitting of sortOrder

The sidebar facets now use the correct sortOrdering (both in jspui and xmlui)

The was an improvement based on the following Jira ticket
https://jira.duraspace.org/browse/DS-263

jonas-atmire committed:
DS-2635 Usage of main div to add results to, fixing empty div problems

aschweer commented:
See Jira issue for my comments. Also, changing the milestone from 5.4 to 6.0 as explained there.

tdonohue commented:
This needs rebasing on the latest master, because of merge conflicts

jonas-atmire commented:
I've rebased this PR on the latest master and set the configuration for "searchFilterContentInOriginalBundle" in discovery.xml that still used the "old" sortOrder

tdonohue commented on dspace/config/modules/workflow.cfg:
```diff
@@ -8,7 +8,7 @@
 #   originalworkflow = Traditional DSpace Workflow
 #   xmlworkflow = New (as of 1.8.0) Configurable Reviewer Workflow
-workflow.framework=originalworkflow
+workflow.workflow.framework=originalworkflow
```
Yuck. Looks like we need to correct the code that is asking for this config, so that the name can be kept as `workflow.framework` instead of prefixing it twice as `workflow.workflow.framework`.

This isn't a fault of your PR. It's just that your PR made me realize we need to fix the code which uses the `workflow.framework` config to avoid this double prefixing.  I'll work on a separate PR to fix this.

tdonohue replied:
This change now needs removal, as DS-3037 is fixed on master.

tdonohue commented:
With this PR, Travis CI keeps killing the processing of the `ITCommunityCollection` integration test:
```
Running org.dspace.content.ITCommunityCollection
Killed
```
This usually implies either that integration test is taking a very long time to complete, or it is taking an inordinate amount of memory or CPU.  I've tried restarting Travis several times, and it always errors out on that same test.

We may want to double check this integration test is still functioning with this PR in place (`mvn clean package -Dmaven.test.skip=false -DskipITs=false`

tdonohue commented:
This is looking good overall, but it needs documentation on *how* to configure these new settings.  That'd be useful also in testing this.

tdonohue commented:
@jonas-atmire : This needs a minor rebase. We've fixed the `workflow.workflow.framework` bug on master (DS-3037), but this PR changes the workflow.cfg to use that double prefix. Please remove that change to workflow.cfg

tdonohue commented:
I also agree with @hardyoyo's note that "view more" on facets should be sorted in *descending* order by default, as that's the same order they are displayed in the sidebar.

jonas-atmire committed:
DS-2635 Default sorting update for both filterPage and sidebar + reverting workflow config change

jonas-atmire commented:
I've rebased the changes onto the latest version of master + reverted the change I'd done to the workflow config file.

I changed the defaults for the "sortOrderFilterPage" for all current to be the same as the order used in the sidebar.
This being sorting on count, descending

terrywbrady commented:
+1 This works well out of the box and will be a great enhancement.  My users have asked for this in the past.

I am attempting to reconstruct some of the custom facets that I have in production.  I am starting a re-index process in order to test those changes.

See the comments from @aschweer in Jira regarding the default behavior of the date facets.  It would be good to add a comment above the sample declaration for the Issue Date indicating that date fields contain their own sort behavior.

It will be important for the DSpace 6 upgrade notes to contain advice on how to upgrade the discovery.xml configuration file.

terrywbrady commented:
I reviewed this feature with our digital services team and they are excited by the ability to paginate through terms by COUNT.  The only concern that was expressed was that a user might want to switch to an alphabetic listing of terms.

I support merging this enhancement as it is.  I think the enhancement would be even more useful with the enhancements suggested by @aschweer.  https://jira.duraspace.org/browse/DS-2635?focusedCommentId=46631&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-46631

hardyoyo commented:
+1, tested, works as expected
